File: backend/src/main/java/com/guardianes/cards/domain/model/Card.java
package com.guardianes.cards.domain.model;

import java.time.LocalDateTime;
import java.util.Objects;

/**
 * Represents a collectible card in the Guardianes de Gaia game. Cards can be collected by scanning
 * QR codes or NFC tags.
 */
public class Card {
  private final Long id;
  private final String name;
  private final String description;
  private final CardElement element;
  private final CardRarity rarity;
  private final Integer attackPower;
  private final Integer defensePower;
  private final Integer energyCost;
  private final String imageUrl;
  private final String qrCode;
  private final String nfcCode;
  private final LocalDateTime createdAt;
  private final boolean isActive;

  public Card(
      Long id,
      String name,
      String description,
      CardElement element,
      CardRarity rarity,
      Integer attackPower,
      Integer defensePower,
      Integer energyCost,
      String imageUrl,
      String qrCode,
      String nfcCode,
      LocalDateTime createdAt,
      boolean isActive) {
    this.id = id;
    this.name = validateName(name);
    this.description = validateDescription(description);
    this.element = Objects.requireNonNull(element, "Element cannot be null");
    this.rarity = Objects.requireNonNull(rarity, "Rarity cannot be null");
    this.attackPower = validateAttackPower(attackPower);
    this.defensePower = validateDefensePower(defensePower);
    this.energyCost = validateEnergyCost(energyCost);
    this.imageUrl = imageUrl;
    this.qrCode = validateQrCode(qrCode);
    this.nfcCode = nfcCode; // NFC code is optional
    this.createdAt = Objects.requireNonNull(createdAt, "Created at cannot be null");
    this.isActive = isActive;
  }

  public static Card create(
      String name,
      String description,
      CardElement element,
      CardRarity rarity,
      Integer attackPower,
      Integer defensePower,
      Integer energyCost,
      String imageUrl,
      String qrCode,
      String nfcCode) {
    return new Card(
        null,
        name,
        description,
        element,
        rarity,
        attackPower,
        defensePower,
        energyCost,
        imageUrl,
        qrCode,
        nfcCode,
        LocalDateTime.now(),
        true);
  }

  // Validation methods
  private String validateName(String name) {
    if (name == null || name.trim().isEmpty()) {
      throw new IllegalArgumentException("Card name cannot be null or empty");
    }
    if (name.length() > 50) {
      throw new IllegalArgumentException("Card name cannot exceed 50 characters");
    }
    return name.trim();
  }

  private String validateDescription(String description) {
    if (description == null || description.trim().isEmpty()) {
      throw new IllegalArgumentException("Card description cannot be null or empty");
    }
    if (description.length() > 500) {
      throw new IllegalArgumentException("Card description cannot exceed 500 characters");
    }
    return description.trim();
  }

  private Integer validateAttackPower(Integer attackPower) {
    if (attackPower == null || attackPower < 0 || attackPower > 999) {
      throw new IllegalArgumentException("Attack power must be between 0 and 999");
    }
    return attackPower;
  }

  private Integer validateDefensePower(Integer defensePower) {
    if (defensePower == null || defensePower < 0 || defensePower > 999) {
      throw new IllegalArgumentException("Defense power must be between 0 and 999");
    }
    return defensePower;
  }

  private Integer validateEnergyCost(Integer energyCost) {
    if (energyCost == null || energyCost < 0 || energyCost > 10) {
      throw new IllegalArgumentException("Energy cost must be between 0 and 10");
    }
    return energyCost;
  }

  private String validateQrCode(String qrCode) {
    if (qrCode == null || qrCode.trim().isEmpty()) {
      throw new IllegalArgumentException("QR code cannot be null or empty");
    }
    if (!qrCode.matches("^[A-Z0-9]{16}$")) {
      throw new IllegalArgumentException("QR code must be 16 alphanumeric characters");
    }
    return qrCode;
  }

  // Business logic methods
  public boolean canBePlayedWith(int availableEnergy) {
    return availableEnergy >= this.energyCost;
  }

  public boolean isElementallyAdvantageousAgainst(Card opponent) {
    return this.element.hasAdvantageOver(opponent.element);
  }

  public int calculateDamageAgainst(Card opponent) {
    int baseDamage = this.attackPower - opponent.defensePower;
    baseDamage = Math.max(baseDamage, 1); // Minimum 1 damage

    if (this.isElementallyAdvantageousAgainst(opponent)) {
      baseDamage = (int) (baseDamage * 1.5); // 50% bonus damage
    }

    return baseDamage;
  }

  public int getTotalPower() {
    return this.attackPower + this.defensePower;
  }

  public boolean isPremium() {
    return this.nfcCode != null && !this.nfcCode.trim().isEmpty();
  }

  // Getters
  public Long getId() {
    return id;
  }

  public String getName() {
    return name;
  }

  public String getDescription() {
    return description;
  }

  public CardElement getElement() {
    return element;
  }

  public CardRarity getRarity() {
    return rarity;
  }

  public Integer getAttackPower() {
    return attackPower;
  }

  public Integer getDefensePower() {
    return defensePower;
  }

  public Integer getEnergyCost() {
    return energyCost;
  }

  public String getImageUrl() {
    return imageUrl;
  }

  public String getQrCode() {
    return qrCode;
  }

  public String getNfcCode() {
    return nfcCode;
  }

  public LocalDateTime getCreatedAt() {
    return createdAt;
  }

  public boolean isActive() {
    return isActive;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Card card = (Card) o;
    return Objects.equals(id, card.id) && Objects.equals(qrCode, card.qrCode);
  }

  @Override
  public int hashCode() {
    return Objects.hash(id, qrCode);
  }

  @Override
  public String toString() {
    return "Card{"
        + "id="
        + id
        + ", name='"
        + name
        + '\''
        + ", element="
        + element
        + ", rarity="
        + rarity
        + ", attackPower="
        + attackPower
        + ", defensePower="
        + defensePower
        + ", energyCost="
        + energyCost
        + ", qrCode='"
        + qrCode
        + '\''
        + ", isPremium="
        + isPremium()
        + '}';
  }
}

File: mobile/guardianes_mobile/lib/features/cards/domain/entities/card.dart
import 'package:equatable/equatable.dart';

enum CardElement { fire, earth, water, air }

enum CardRarity { common, uncommon, rare, epic, legendary }

class Card extends Equatable {
  final int id;
  final String name;
  final String description;
  final CardElement element;
  final CardRarity rarity;
  final int attackPower;
  final int defensePower;
  final int energyCost;
  final String? imageUrl;
  final String qrCode;
  final String? nfcCode;
  final DateTime createdAt;
  final bool isActive;

  const Card({
    required this.id,
    required this.name,
    required this.description,
    required this.element,
    required this.rarity,
    required this.attackPower,
    required this.defensePower,
    required this.energyCost,
    this.imageUrl,
    required this.qrCode,
    this.nfcCode,
    required this.createdAt,
    required this.isActive,
  });

  bool get isPremium => nfcCode != null && nfcCode!.isNotEmpty;

  int get totalPower => attackPower + defensePower;

  bool canBePlayedWith(int availableEnergy) {
    return availableEnergy >= energyCost;
  }

  bool hasElementalAdvantageOver(Card opponent) {
    const advantageMap = {
      CardElement.fire: CardElement.earth,
      CardElement.earth: CardElement.water,
      CardElement.water: CardElement.air,
      CardElement.air: CardElement.fire,
    };
    return advantageMap[element] == opponent.element;
  }

  int calculateDamageAgainst(Card opponent) {
    int baseDamage = attackPower - opponent.defensePower;
    baseDamage = baseDamage < 1 ? 1 : baseDamage; // Minimum 1 damage

    if (hasElementalAdvantageOver(opponent)) {
      baseDamage = (baseDamage * 1.5).round(); // 50% bonus damage
    }

    return baseDamage;
  }

  @override
  List<Object?> get props => [
        id,
        name,
        element,
        rarity,
        attackPower,
        defensePower,
        energyCost,
        qrCode,
        isActive,
      ];
}

extension CardElementExtension on CardElement {
  String get displayName {
    switch (this) {
      case CardElement.fire:
        return 'Fuego';
      case CardElement.earth:
        return 'Tierra';
      case CardElement.water:
        return 'Agua';
      case CardElement.air:
        return 'Aire';
    }
  }

  String get emoji {
    switch (this) {
      case CardElement.fire:
        return '🔥';
      case CardElement.earth:
        return '🌍';
      case CardElement.water:
        return '💧';
      case CardElement.air:
        return '🌪️';
    }
  }
}

extension CardRarityExtension on CardRarity {
  String get displayName {
    switch (this) {
      case CardRarity.common:
        return 'Común';
      case CardRarity.uncommon:
        return 'Poco Común';
      case CardRarity.rare:
        return 'Raro';
      case CardRarity.epic:
        return 'Épico';
      case CardRarity.legendary:
        return 'Legendario';
    }
  }

  String get colorEmoji {
    switch (this) {
      case CardRarity.common:
        return '⚪';
      case CardRarity.uncommon:
        return '🟢';
      case CardRarity.rare:
        return '🔵';
      case CardRarity.epic:
        return '🟣';
      case CardRarity.legendary:
        return '🟡';
    }
  }

  int get experiencePoints {
    switch (this) {
      case CardRarity.common:
        return 10;
      case CardRarity.uncommon:
        return 25;
      case CardRarity.rare:
        return 50;
      case CardRarity.epic:
        return 100;
      case CardRarity.legendary:
        return 200;
    }
  }

  int get tradeValue {
    switch (this) {
      case CardRarity.common:
        return 1;
      case CardRarity.uncommon:
        return 3;
      case CardRarity.rare:
        return 10;
      case CardRarity.epic:
        return 25;
      case CardRarity.legendary:
        return 100;
    }
  }
}
